### PR TITLE
feat: add wait function for custom expectations

### LINKED ISSUE #61: Add waitForElementToDiminish to helper functions
I started using the library in our react-native project and I keep facing problems to wait until ApolloClient finish loading. Why don't we implement a waitForElementToDiminish so we can always add the loading label/testID and simplify our waits.

### PR DESCRIPTION
**Pull request by flpStrri**
### Summary
Closes #61 adding a wait function that gives the library user more autonomy writing a complex expectation and not simply waiting for a component.

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -247,6 +247,32 @@ test('waiting for an Banana to be ready', async () => {
...
+Waits for non-deterministic periods of time until your expectation function succeed or times out. `wait` is a small wrapper
+around the [`wait-for-expect`](https://github.com/TheBrainFamily/wait-for-expect) module.
+
+```jsx
+import { render, wait } from 'react-testing-library';
+
+test('waiting for an Banana to be ready', async () => {
+  const { getByText, toJSON } = render(<Banana />);
+
+  await wait(expect(getByText('Loading...')..toThrow()));
```
Comment (thymikee) [inline docs/api.md]: I believe this won't work, because `wait` expects returning a promise, and neither `expect` nor `getByText` does that.

File: package.json [hunk 2]
```diff
@@ -33,7 +33,8 @@
     "typescript": "^3.1.1"
   },
   "dependencies": {
-    "pretty-format": "^23.6.0"
+    "pretty-format": "^23.6.0",
+    "wait-for-expect": "^1.1.0"
```
Comment (thymikee) [inline package.json]: please remove this dep, `waitForElement` does basically the same

File: src/__tests__/wait.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { wait } from '..';
```
Comment (thymikee) [inline src/__tests__/wait.js]: missing `// @flow` pragma

File: src/wait.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// @flow
+import waitForExpect from 'wait-for-expect';
+
+export default function wait(
+  expectation: () => Promise<{}> = () => Promise.resolve({}),
+  timeout: number = 4500,
+  interval: number = 50
+): Promise<{}> {
+  return waitForExpect(expectation, timeout, interval);
```
Comment (thymikee) [inline src/wait.js]: you can use `waitForElement` here, it should work the same.

File: typings/index.d.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
-import * as React from 'react';
-import { ReactTestInstance, ReactTestRendererJSON } from 'react-test-renderer';
+import * as React from "react";
+import { ReactTestInstance, ReactTestRendererJSON } from "react-test-renderer";
```
Comment (thymikee) [inline typings/index.d.ts]: please revert these changes

**Comment by flpStrri** [comment]
I understood what you said, makes total sense. `wait()` is not solving any special problem other than the default resolve promise. I am just not 100% happy with how it reads out:

```javascript
test('waitForElement expect test', async() => {
  let a = 1
  setTimeout(() => {
    a = 2
  }, 30)
  expect(a).toEqual(1)
  await waitForElement(() => expect(a).toEqual(2))
  expect(a).toEqual(2)
})
```

**Comment by flpStrri** [comment]
I don't mind making the needed changes for that. I will work on it. :D

**Comment by thymikee** [comment]
Ok, let's go with `wait` as an alias with default expectation :)

**Comment by flpStrri** [comment]
@thymikee, can you give it a look if it is working now?

**Comment by thymikee** [comment]
Just re-triggering the CI, because it was not building from forks and I just fixed it

File: docs/api.md [hunk 6]
```diff
@@ -247,6 +247,33 @@ test('waiting for an Banana to be ready', async () => {
 });
 ```
 
+## `wait`
+
+Defined as:
+
+```jsx
+function wait(
+  expectation: () => Promise<{}> = () => Promise.resolve({}),
```
Comment (thymikee) [inline docs/api.md]:
```suggestion
  expectation?: () => Promise<{}>,
```

File: docs/api.md [hunk 7]
```diff
@@ -247,6 +247,33 @@ test('waiting for an Banana to be ready', async () => {
...
+of the `waitForElement` function with a default empty expectation that can be used to make a one tick wait.
+
+```jsx
+import { render, wait } from 'react-testing-library';
+
+test('waiting for an Banana to be ready', async () => {
+  const { getByText, toJSON } = render(<Banana />);
+
+  const loadingScreenHasDisappearedTest = () => expect(getByText('Loading...')).toThrow();
+  await wait(loadingScreenHasDisappearedTest);
```
Comment (thymikee) [inline docs/api.md]: this example test doesn't really make sense, because we literally wait for element, not some arbitrary async operation to complete.

File: docs/api.md [hunk 8]
```diff
@@ -247,6 +247,33 @@ test('waiting for an Banana to be ready', async () => {
 });
 ```
 
+## `wait`
+
+Defined as:
```
Comment (thymikee) [inline docs/api.md]: Let's just state that this is a convenience sugar over `waitForElement` to let you wait for any async value to get present, not only elements. Plus it's possible to run it without an argument to just await next event loop tick

**Comment by Peretz30** [comment]
any updates?
